crates/extensions/c8y_mapper_ext/src/availability/builder.rs

```diff
@@ -0,0 +1,151 @@
+use tedge_actors::DynSender;
+use tedge_actors::LoggingSender;
+use tedge_actors::MessageSink;
+use tedge_actors::MessageSource;
+use tedge_actors::NoConfig;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::RuntimeRequestSink;
+use tedge_actors::Service;
+use tedge_actors::SimpleMessageBoxBuilder;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::ChannelFilter;
+use tedge_api::mqtt_topics::EntityFilter;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::HealthStatus;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+use tedge_mqtt_ext::TopicFilter;
+
+pub struct AvailabilityBuilder {
+    config: AvailabilityConfig,
+    box_builder: SimpleMessageBoxBuilder<AvailabilityInput, AvailabilityOutput>,
+    timer_sender: DynSender<TimerStart>,
+}
+
+impl AvailabilityBuilder {
+    pub fn new(
+        config: AvailabilityConfig,
+        mqtt: &mut (impl MessageSource<MqttMessage, TopicFilter> + MessageSink<MqttMessage>),
+        timer: &mut impl Service<TimerStart, TimerComplete>,
+    ) -> Self {
+        let mut box_builder: SimpleMessageBoxBuilder<AvailabilityInput, AvailabilityOutput> =
+            SimpleMessageBoxBuilder::new("AvailabilityMonitoring", 16);
+
+        box_builder.connect_mapped_source(
+            Self::subscriptions(&config.mqtt_schema),
+            mqtt,
+            Self::mqtt_message_parser(config.clone()),
+        );
+
+        mqtt.connect_mapped_source(
+            NoConfig,
+            &mut box_builder,
+            Self::mqtt_message_builder(config.clone()),
+        );
+
+        let timer_sender = timer.connect_client(box_builder.get_sender().sender_clone());
+
+        Self {
+            config: config.clone(),
+            box_builder,
+            timer_sender,
+        }
+    }
+
+    fn subscriptions(mqtt_schema: &MqttSchema) -> TopicFilter {
+        [
+            mqtt_schema.topics(EntityFilter::AnyEntity, ChannelFilter::EntityMetadata),
+            mqtt_schema.topics(EntityFilter::AnyEntity, ChannelFilter::Health),
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    fn mqtt_message_parser(
+        config: AvailabilityConfig,
+    ) -> impl Fn(MqttMessage) -> Option<AvailabilityInput> {
+        move |message| {
+            if let Ok((source, channel)) = config.mqtt_schema.entity_channel_of(&message.topic) {
+                match channel {
+                    Channel::EntityMetadata => {
+                        if let Ok(registration_message) =
+                            EntityRegistrationMessage::try_from(&message)
+                        {
+                            return Some(registration_message.into());
+                        }
+                    }
+                    Channel::Health => {
+                        // Support only the default service schema: "device/+/service/+/status/health"
+                        if source.is_default_service() {
```

Comment:
Why this restriction? It seems to be unused, at least here.

Reply:
This is actually used.
- `EntityTopicId` can be anything as long as it has 4 segments. `custom/foo/bar/baz` is a valid `EntityTopicId`.
- `mqtt_schema.entity_channel_of(&topic)` checks if the topic is `EntityTopicId` + `/status/health`.

So, if `source.is_default_servie()` gets rid of something like the `source` = `custom/foo/bar/baz/status/health`.

Or is your question why it doesn't accept `custom/foo/bar/baz/status/health`?

Reply:
> Or is your question why it doesn't accept `custom/foo/bar/baz/status/health`?

Yes, why `te/custom/foo/bar/baz/status/health` should be rejected?
Why not accept any `te/+/+/+/+/status/health` messages?

Reply:
Done in cb41d2df39dd6dfb81f47ee784af77e08368e2f6